### DISCUSSION TITLE
VBA rename a OneNote page (change the page title)

### DISCUSSION THREAD

**Discussion by plafontaine** [community/community#126617] [Programming Help] [answered]
### Body

Hi,
     I would like to change a page title using vba in onenote. I m able to create new page but I just cant find how to rename the page (updatehierarchy not working). I found a great code here: https://github.com/pierrms/Alphabetize-OneNote-Sections-VBA/blob/master/OneNote_Sort_VBA.bas

Its working well exept its not adapted to change page name. Any help would be appreciated, pretty sure the code only needs few lines to includes pages nor remove all relatives to notebooks and sections and only includes pages stuffs...

Thanks !

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by mranv** [comment]
To rename a OneNote page using VBA, you need to modify the OneNote page XML and then use the `UpdateHierarchy` method to apply the changes. Here's a step-by-step guide to help you achieve this:

1. **Access the OneNote Application Object:**
   To work with OneNote from VBA, you need to create an instance of the OneNote application.

2. **Retrieve the Page's XML:**
   You need to get the current XML of the page you want to rename.

3. **Modify the Page Title:**
   Once you have the XML, you can modify the title element.

4. **Update the Page with the New XML:**
   Finally, use the `UpdateHierarchy` method to update the page with the new XML.

Here's an example of how you can achieve this:

```vba
Sub RenameOneNotePage()

    ' Variables
    Dim oneNoteApp As Object
    Dim ns As Object
    Dim sectionID As String
    Dim pageID As String
    Dim newTitle As String
    Dim pageXML As String
    Dim updatedPageXML As String
    
    ' Set new page title
    newTitle = "New Page Title"
    
    ' Initialize OneNote Application object
    Set oneNoteApp = CreateObject("OneNote.Application")
    
    ' Retrieve the ID of the page you want to rename
    ' You might need to navigate through the notebooks and sections to find the specific pageID
    ' Here we assume you already have the pageID and sectionID

    pageID = "YOUR_PAGE_ID_HERE"
    sectionID = "YOUR_SECTION_ID_HERE"
    
    ' Retrieve the XML of the page
    oneNoteApp.GetPageContent pageID, pageXML, 0
    
    ' Load the XML into MSXML2.DOMDocument for parsing
    Dim xmlDoc As Object
    Set xmlDoc = CreateObject("MSXML2.DOMDocument")
    xmlDoc.LoadXML pageXML
    
    ' Change the title in the XML
    xmlDoc.SelectSingleNode("//one:Title/one:Text").Text = newTitle
    
    ' Get the updated XML
    updatedPageXML = xmlDoc.XML
    
    ' Update the page with the new XML
    oneNoteApp.UpdatePageContent updatedPageXML
    
    MsgBox "Page title updated to: " & newTitle

End Sub
```

### Key Points:
1. **Initialize OneNote Application Object:**
   ```vba
   Set oneNoteApp = CreateObject("OneNote.Application")
   ```

2. **Retrieve Page Content:**
   ```vba
   oneNoteApp.GetPageContent pageID, pageXML, 0
   ```

3. **Load and Modify XML:**
   ```vba
   Dim xmlDoc As Object
   Set xmlDoc = CreateObject("MSXML2.DOMDocument")
   xmlDoc.LoadXML pageXML
   xmlDoc.SelectSingleNode("//one:Title/one:Text").Text = newTitle
   updatedPageXML = xmlDoc.XML
   ```

4. **Update Page Content:**
   ```vba
   oneNoteApp.UpdatePageContent updatedPageXML
   ```

### Notes:
- You need to know the `pageID` of the page you want to rename.
- You might need to navigate through the notebooks and sections to find the specific page if you don't have the `pageID`.
- Make sure the OneNote namespace (`one:Title/one:Text`) matches the actual XML structure of the OneNote page.

This should help you rename a OneNote page using VBA. If you need further customization or assistance with navigating the notebooks and sections, feel free to ask!

**Reply by plafontaine** [reply, marked as answer]
Hi, thanks for this! I just try it the same way you wrote it but  I had nearly the same thing exept that every time I try to selectsinglenode() with title, the variable stay 'nothing' so not able to update anything. Looks like I just have a page name, when I update the page name, I cannot update the xml... it gives me an error...

`onenote.UpdateHierarchy doc.XML ` works but do nothing , like it just update the old xml 

`nodepage.Attributes.getNamedItem("name").Text = "allo"`  to set the page name (nodepage is the page) because as I said selectsinglenode do not works and when I use msgbox nodepage.xml I just see page name, id... but no title even if I have one on the page.

so: `xmlDoc.SelectSingleNode("//one:Title/one:Text").Text = newTitle` give an error (variable not defined) probably because of cannot selectsinglenode and `oneNoteApp.UpdatePageContent updatedPageXML` gives automation error.

Do you have any ideas on this?
thanks!

**Comment by plafontaine** [comment]
The code started to work suddenly... strange but thanks !
